### ISSUE TITLE
DDG improvements

### ISSUE THREAD

**Issue by axt** [angr/angr#224]
I'm trying to understand the ddg.py analysis, and found multiple bugs in it.

My questions are:
- is it under internal changes, or may I send a pull-request with the fixes?
- should we discuss the bugs/changes here or with the pull request?

**Comment by ltfish** [comment]
Give me a PR!


> should we discuss the bugs/changes here or with the pull request?

Either way.

**Comment by axt** [comment]
@ltfish Will send a PR with the fixes. Expect it in a few days. I propose to let this issue open till I send the PR for related discussions.

**Comment by ltfish** [comment]
Sure. No need to hurry!

**Comment by axt** [comment]
I already have a question.

For SimRegisterVariables the size attribute is specified in bits.
For SimStackVariable, the size represent bytes.
```
    def __repr__(self):
            s = "<Stack %s%s, %s bytes>" % (self.base, offset, size)
```
I assume that for SimMemoryVariable which is the parent of SimStackVariable, the size also should represent bytes.

Is those assumptions correct?
Is that on purpose?

**Comment by ltfish** [comment]
Initially the `size` field started as bits, but now I believe they should *all* be bytes.

**Comment by axt** [comment]
Thanks. 
Checking simuvex use-cases, SimRegisterVariables are used with size representing bytes. Will update DDG and VSA_DDG according to this then.

**Comment by ltfish** [comment]
DDG was definitely written and re-written in hassles :-)
Thanks!

**Comment by axt** [comment]
Found something interesting!

'Add' and 'Sub' operations are handled in `_track`. The relevant code for 'Add':
```
[1]    sort, offset = temp_register_symbols[tmp]
[2]    offset += expr_1.ast.args[0]
       data_generated = (sort, offset)
```
Currently at [1] offsets can be negative and positive, lets assume offset is -4. The code does an (Add, 0xfffffff8), so at [2] offset becomes 0xfffffff4, instead of -0xc. The problem with this is that while both (sp+0xfffffff4 and sp-0xc) represent the same memory location, but two different SimMemoryVariables will be created for them.

Although I didn't see any problem arising from this, I think its still a logical error.

To demonstrate this, check the SimStackVariables with the original code, and with some correction:
Current code:
```
-0x8001004c set([<0x804840e 0x804840e(3)>])
-0x8001004b set([<0x804840e 0x804840e(3)>])
-0x8001004a set([<0x804840e 0x804840e(3)>])
-0x80010049 set([<0x804840e 0x804840e(3)>])
0x7ffeffb8 set([<0x80483ed 0x80483ed(5)>])
0x7ffeffb9 set([<0x80483ed 0x80483ed(5)>])
0x7ffeffba set([<0x80483ed 0x80483ed(5)>])
0x7ffeffbb set([<0x80483ed 0x80483ed(5)>])
``` 

Fixed up code:
```
0x7ffeffb4 set([<0x804840e 0x804840e(3)>])
0x7ffeffb5 set([<0x804840e 0x804840e(3)>])
0x7ffeffb6 set([<0x804840e 0x804840e(3)>])
0x7ffeffb7 set([<0x804840e 0x804840e(3)>])
0x7ffeffb8 set([<0x80483ed 0x80483ed(5)>])
0x7ffeffb9 set([<0x80483ed 0x80483ed(5)>])
0x7ffeffba set([<0x80483ed 0x80483ed(5)>])
0x7ffeffbb set([<0x80483ed 0x80483ed(5)>])
```

What do you think about this?

**Comment by ltfish** [comment]
You are definitely correct. The problem didn't arise, since VEX almost always generates IRs like `Add(sp, 0xfffffff4)`, other than its counterpart with a negative offset.

**Comment by axt** [comment]
Any idea of how/where should be this fixed?

**Comment by ltfish** [comment]
I think it should be fixed in DDG itself. Regarding _how_, not at this moment :-(

I'm open to ideas/suggestions!

Feel free to leave it as an issue for me. You can open a new one.

**Comment by rhelmot** [comment]
One solution would be to do all the arithmetic with bitvectors - a recent change to claripy made it so that all BVVs are normalized to unsigned values on construction.

**Comment by ltfish** [comment]
@rhelmot: that should work, but being unable to use `==` on claripy ASTs is really annoying.

**Comment by axt** [comment]
@ltfish Sent a pull-request with some minor changes. Please have a look at it.

In order to make it fully functional, two issues remain:
- https://github.com/angr/angr/issues/223
- the address normalization, i think its better if I leave it to you (shoud I open a separate ticket for it)

**Comment by axt** [comment]
So I've made a simple example:
```
int c2(int a) { return a; }
int c1(int a) { if (a > 0) c2(a); else c2(-a); }
int c0(int a) { c1(a); }
int main() { int a; c0(a); }
```
And I've plotted all the `node -> successing_node` "transitions" to better understand what the code does:
```diff
                changed = False

                for successing_node in successing_nodes:
+                   self._track_history.append((node, successing_node))
                    if (state.scratch.jumpkind == 'Ijk_Call' or state.scratch.jumpkind.startswith('Ijk_Sys')) and \
```
Labels on the edges indicate the order in which they were executed.

I think these are the bugs:
- it follows the fake rets (e.g: 0,3,10,12), even if there is another successors of the node
- there are some doubed edges (e.g:6,7,8,9), because we execute every successing node for every final state
- the processing order is not optimal (e.g: 2)

I think the first one can introduce some extra edges in the graph, the others are just making it do more work, than necessary.

![ddg_old](https://cloud.githubusercontent.com/assets/759365/19023104/50d896ca-88e6-11e6-944c-caf8b5c98c1b.png)

**Comment by ltfish** [comment]
> it follows the fake rets (e.g: 0,3,10,12), even if there is another successors of the node

I agree this is a bug - I thought I have logics handling that.

The other two issues can be resolved by introducing a better graph traversal logic. Are you working on the DDG code? I want to refactor it a little bit to improve the readability.

It's awesome to work with someone who wants to read my crazy code :-)

**Comment by axt** [comment]
Yes, there is some logic handling that, I didn't checked why its not working. If we would fix this, the graph traversal order will get better by just that fix.

Not working on the DDG right now, because I got distracted a bit :-), and trying to do some simple signed/unsigned type inference analysis based on the DDG right now. So feel free to do the refactor, any  improvements of it helps my case too, and because I depend on it, I will do a lot of testing.

Not sure what are your plans regarding the refactor, may I share some ideas? :)
- the working queue could be extracted into a separate class, maybe even we could use ForwardAnalysis instead of it (? not sure about that)
- Instead of the current calldepth handling we could do something what is in binaryoptimizer, to rebuild the cfg with basegraph and calldepth parameters set

I like working with you too, and I would glad to help anyway I can, but at least with the testing, only that my time is limited what I can spend on this.

**Comment by axt** [comment]
Oh, and I can ask a lot of questions ...

I found that every time we introduce a variable (initial=true),  we have a node with the same location, but with a SimTempVariable as its successor, and initial=false. 

I couldn't decide if this is intentional or maybe a bug. Can you share your opinion about this?

![ddg_initnode](https://cloud.githubusercontent.com/assets/759365/19208414/d347cdf8-8cf9-11e6-9a53-2f16c491f437.png)

**Comment by ltfish** [comment]
> the working queue could be extracted into a separate class, maybe even we could use ForwardAnalysis instead of it (? not sure about that)

+1

> Instead of the current calldepth handling we could do something what is in binaryoptimizer, to rebuild the cfg with basegraph and calldepth parameters set

I'm not sure if I understand it. Can you please elaborate on this?

I'll split the huge working queue into different functions at the very least. It's a giant mess right now. Also I'll introduce a better traversal order since we already have access to the CFG at this moment.

**Comment by ltfish** [comment]
> I found that every time we introduce a variable (initial=true), we have a node with the same location, but with a SimTempVariable as its successor, and initial=false. 

This is expected. What you see here comes from the IR statement `t153 = LDle:I32(t151)`, which creates two different `SimAction`s:
- loading from memory location `t151`, and
- writing the value to `t153`.

`initial` implies "a new variable, probably an input value with respect to this block, the current function, or whatever we are analyzing, is introduced". I guess during the analysis, we've never read from the location that `t151` points to, that's why it introduces a new `SimMemoryVariable` with `initial` being True. Then the value is written to a `SimTemporaryVariable`, at which moment we are not introducing a new value, but assigning an existing value to `t153`.

**Comment by axt** [comment]
Thanks for the explanation, I see now that it is correct!

> I'm not sure if I understand it. Can you please elaborate on this?

The problems arise when we reach the `call_depth` specified. 

Lets look at this simple example:
```
A ---call---> B ---ret---> C
 \--------fakeret---------^
```
If we have reached the call depth, then we don't want to trace into B, instead we want to continue tracing with node C. 

This code was part of the DDG once (now commented out, but still used in VSA_DDG) to select the successors to continue the tracing with.
```python
corresponding_successors = [n for n in successing_nodes if not state.ip.symbolic and n.addr == state.se.any_int(state.ip)]
```
The problem with this is, that in the case above  C will not be in corresponding_successors, since A's final state "points to" B.

What the code does currently is to follow all existing successors for every final state. This is also not correct but it helps in this case, since we will follow the fakeret edge, so tracing into C will happen. As we agreed following the fakeret edge is usually a bug, but when the call depth is reached, actually this is what we should do.

**So I think what we need is the clever combination of this two logics**

What I thought (originally) when I wrote the previous comment (**I see now that this is not needed!**):
In `binary_optimizer.py` the analysis is initialized with a cfg. Then later for every function it builds a new cfg like this (call_depth=1, +using the original cfg/kb as a base):
```python
cfg = self.project.analyses.CFGAccurate(kb=func_kb, call_depth=1,base_graph=function.graph,keep_state=True,starts=(function.addr,),iropt_level=0)
```
So I thought we could "rebuild" the CFG, with the corresponding `call_depth` parameter, using the original CFG as a base. Then we could use this rebuilt CFG to help the traversal.

**Comment by ltfish** [comment]
> So I thought we could "rebuild" the CFG, with the corresponding `call_depth` parameter, using the original CFG as a base.

I think it suffices to pass in a `CFGAccurate` instance with the corresponding `call_depth`.

**Comment by axt** [comment]
I agree.

Found another interesting thing:
```python
for s,t,data in ddg.data_graph.in_edges(n, data=True):
    print s
```
Yields: `<Stack bp+0x8, 4 bytes> `
vs 
```python
for s,t,data in ddg.data_graph.edges(data=True):
  if t != n:
    continue
  print s
```
Yields: `<Stack sp+0x4, 4 bytes> `

**Comment by axt** [comment]
I have an idea for this: Both `<Stack sp+0x4, 4 bytes>` and `<Stack bp+0x8, 4 bytes>` has the same `addr` property, with different base and offset. The hash code of those two objects will be generated based only on `addr` (which is right in my opinion), but they share the same cache-key. Not really sure why the two iteration find the different instances, but I think its related to this.

**Comment by mattrepl** [comment]
Hey @axt, saw this and your related discussion #242. Any more work being done on this? I'm especially interested in getting calling convention detection improved. 

There's a reference to using DDG for analyzing function I/O in `cfg_accurate.py`. Any idea if I should be helping on improving DDG generation if I want calling convention matching to be improved? And is there any interim solution I should consider while DDG is being improved?

Currently, `CFGAccurate` tries to identify argument stack variables and registers but it appears to wildly over approximate. There's a refinement process, `_refine_function_arguments`, in `CFGAccurate` I'm still working through the code; however, in practice the current method fails to identify the cdecl calling convention used in simple x86 programs.

**Comment by axt** [comment]
Hi @mattrepl 

There are some problems with current DDG generation, what needs to be solved:

- These two SimStackVariables have the same hash: `<Stack sp+0x4, 4 bytes>`, `<Stack bp+0x8, 4 bytes>`, which can make digraph edge iterations inconsistent, see https://github.com/networkx/networkx/issues/2282

- Normalization and keep-state are not consistent in the current version: https://github.com/angr/angr/issues/223

- Other ones mentioned in this thread.

As far as I know, @ltfish is working on some improvements, which probably will affect DDG generation too. 

I've shared the code I've done so far, but need to warn you, it is in a pretty messy state currently.

You can check it here: https://github.com/axt/angr-ti

CFG and DDG visualization is depending on https://github.com/axt/bingraphvis, you need to install that, or remove the calls to graph generation. 

Your contribution and ideas are welcome, we can continue discussion here: https://github.com/axt/angr-ti/issues/1

**Comment by ltfish** [comment]
@mattrepl Axt is right that there is a huge improvement branch being discussed internally and waiting to be merged to master soon. During the past week, there was a disease outbreak here in the lab, and all progresses have been slowed down a bit. Once that branch is merged in, DDG refactoring will also be merged in within a few days.

Calling convention detection is something I have worked on, and have some code lying around in another internal project. I will import the code to angr once DDG refactoring is done.

**Comment by mattrepl** [comment]
Thanks for the info. Sounds like I'll hold off for the internal improvements to be merged here and review axt's work in the meantime.

Has there been any thought in moving general internal development to this repo? I realize there will likely be angr components that need to be kept internal until papers are published, but would definitely improve collaboration if we can all work from the same codebase.

**Comment by ltfish** [comment]
> Has there been any thought in moving general internal development to this repo?

Yes, we will move angr development to GitHub. @zardus is taking care of it. It won't be done in a few weeks, but that's definitely the eventual goal.

> I realize there will likely be angr components that need to be kept internal until papers are published

The idea is that any private components should be written as an external module and then being plugged in to angr. Right now all of our on-going "secret" projects are already done in this way. Making angr flexible is our top priority.

**Comment by axt** [comment]
Since DDG will be reimplemented I close this issue.
